cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetTypeService.java

```diff
@@ -0,0 +1,417 @@
+import co.cask.cdap.proto.id.DatasetModuleId;
+import co.cask.cdap.proto.id.DatasetTypeId;
+import co.cask.cdap.proto.id.EntityId;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.security.Action;
+import co.cask.cdap.proto.security.Principal;
+import co.cask.cdap.security.authorization.AuthorizerInstantiator;
+import co.cask.cdap.security.spi.authentication.AuthenticationContext;
+import co.cask.cdap.security.spi.authorization.AuthorizationEnforcer;
+import co.cask.cdap.security.spi.authorization.Authorizer;
+import co.cask.cdap.security.spi.authorization.UnauthorizedException;
+import co.cask.http.BodyConsumer;
+import co.cask.http.HttpResponder;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
+import com.google.common.io.Files;
+import com.google.inject.Inject;
+import org.apache.twill.filesystem.Location;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Manages lifecycle of dataset {@link DatasetType types} and {@link DatasetModule modules}.
+ */
+public class DatasetTypeService {
+  private static final Logger LOG = LoggerFactory.getLogger(DatasetTypeService.class);
+
+  private final DatasetTypeManager typeManager;
+  private final NamespaceQueryAdmin namespaceQueryAdmin;
+  private final NamespacedLocationFactory namespacedLocationFactory;
+  private final AuthorizationEnforcer authorizationEnforcer;
+  private final Authorizer authorizer;
+  private final AuthenticationContext authenticationContext;
+  private final CConfiguration cConf;
+
+  @Inject
+  @VisibleForTesting
+  public DatasetTypeService(DatasetTypeManager typeManager, NamespaceQueryAdmin namespaceQueryAdmin,
+                            NamespacedLocationFactory namespacedLocationFactory,
+                            AuthorizationEnforcer authorizationEnforcer,
+                            AuthorizerInstantiator authorizerInstantiator, AuthenticationContext authenticationContext,
+                            CConfiguration cConf) {
+    this.typeManager = typeManager;
+    this.namespaceQueryAdmin = namespaceQueryAdmin;
+    this.namespacedLocationFactory = namespacedLocationFactory;
+    this.authorizationEnforcer = authorizationEnforcer;
+    this.authorizer = authorizerInstantiator.get();
+    this.authenticationContext = authenticationContext;
+    this.cConf = cConf;
+  }
+
+  /**
+   * Returns all {@link DatasetModuleMeta dataset modules} in the specified {@link NamespaceId namespace}.
+   */
+  List<DatasetModuleMeta> listModules(final NamespaceId namespaceId) throws Exception {
+    ensureNamespaceExists(namespaceId);
```

Comment:
All the code in this class other than the authorization code simply moved from ``DatasetTypeHandler``